plugins/org.eclipse.fordiac.ide.export.forte_ng/src/org/eclipse/fordiac/ide/export/forte_ng/ForteFBTemplate.xtend

```diff
@@ -377,20 +379,35 @@ abstract class ForteFBTemplate<T extends FBType> extends ForteLibraryElementTemp
 
 	def protected generateSetInitialValuesDeclaration(Iterable<VarDeclaration> variables) '''
 		«IF !variables.empty»
-			void setInitialValues() override;
+			void setInitialValues(bool retain) override;
 		«ENDIF»
 	'''
 
 	def protected generateSetInitialValuesDefinition(Iterable<VarDeclaration> variables) '''
 		«IF !variables.empty»
-			void «className»::setInitialValues() {
-			  «FOR variable : variables»
-			  	«variable.generateName» = «variable.generateVariableDefaultValue»;
-			  «ENDFOR»
+			void «className»::setInitialValues(bool retain) {
+				«IF variables.exists[ it.getAttributeValue(LibraryElementTags.RETAIN_ATTRIBUTE) !== null ]»
+					«generateRetainedInitialValuesDefinition(variables)»
+				«ELSE»
+					«generateVariables(variables)»
+				«ENDIF»
 			}
 			
 		«ENDIF»
 	'''
+	def private generateRetainedInitialValuesDefinition(Iterable<VarDeclaration> variables) '''
+		//values that should be retained
+		if (!retain) {
+			«generateVariables(variables.filter[it.getAttributeValue(LibraryElementTags.RETAIN_ATTRIBUTE) !== null && it.getAttributeValue(LibraryElementTags.RETAIN_ATTRIBUTE).equals(RetainTag.RETAIN.string)])»
+		}
+		«generateVariables(variables.filter[it.getAttributeValue(LibraryElementTags.RETAIN_ATTRIBUTE) === null || !it.getAttributeValue(LibraryElementTags.RETAIN_ATTRIBUTE).equals(RetainTag.RETAIN.string)])»
+	'''
+	
+	def private generateVariables(Iterable<VarDeclaration> variables)'''
```

Comment:
Please give this method a more descriptive name, such as `generateVariableDefaultAssignments`.